src/visualization/src/CameraAnimator.cpp

```diff
@@ -0,0 +1,332 @@
+        m_wheelMoving = true;
+        m_wheelDirection = event.MouseInput.Wheel;
+        break;
+
+    case irr::EMIE_LMOUSE_DOUBLE_CLICK:
+    case irr::EMIE_RMOUSE_DOUBLE_CLICK:
+    case irr::EMIE_MMOUSE_DOUBLE_CLICK:
+    case irr::EMIE_LMOUSE_TRIPLE_CLICK:
+    case irr::EMIE_RMOUSE_TRIPLE_CLICK:
+    case irr::EMIE_MMOUSE_TRIPLE_CLICK:
+    case irr::EMIE_COUNT:
+        return false;
+    }
+    return true;
+}
+
+
+//! OnAnimate() is called just before rendering the whole scene.
+void CameraAnimator::animateNode(irr::scene::ISceneNode *node, irr::u32 /*timeMs*/)
+{
+    if (!node || node->getType() != irr::scene::ESNT_CAMERA)
+        return;
+
+    irr::scene::ICameraSceneNode* camera = static_cast<irr::scene::ICameraSceneNode*>(node);
+
+    // If the camera isn't the active camera, and receiving input, then don't process it.
+    if (!camera->isInputReceiverEnabled())
+        return;
+
+    irr::scene::ISceneManager * smgr = camera->getSceneManager();
+    if (smgr && smgr->getActiveCamera() != camera)
+        return;
+
+    irr::core::vector2df mouseDelta = m_mousePos - m_initialMousePosition;
+
+    if (m_wheelMoving)
+    {
+        m_wheelMoving = false;
+        m_zooming = true;
+    }
+
+    if (m_mouseKeys[0])
+    {
+        m_rotating = true;
+        m_initialMousePosition = m_mousePos;
+    }
+
+    if (m_mouseKeys[1])
+    {
+        m_movingUp = true;
+        m_initialMousePosition = m_mousePos;
+    }
+
+    if (m_mouseKeys[2])
+    {
+        m_translating = true;
+        m_initialMousePosition = m_mousePos;
+    }
+
+    if (!m_zooming && !m_translating && !m_rotating && !m_movingUp) //Doing nothing
+    {
+        m_cameraAxis->setVisible(false);
+        return;
+    }
+
+    irr::core::vector3df initialTarget = camera->getTarget();
+    irr::core::vector3df newTarget = initialTarget;
+
+    irr::core::vector3df initialPosition = camera->getPosition();
+    irr::core::vector3df newPosition = initialPosition;
+    irr::core::vector3df upVector = camera->getUpVector();
+
+    const irr::scene::SViewFrustum* const va = camera->getViewFrustum();
+
+    irr::core::matrix4 initialTransformation;
+    initialTransformation.setTranslation(initialPosition);
+
+    irr::core::vector3df xAxis = va->getFarLeftUp() - va->getFarLeftDown();
+    xAxis.normalize();
+    irr::core::vector3df yAxis = va->getFarRightDown() - va->getFarLeftDown();
+    yAxis.normalize();
+
+    irr::core::vector3df zAxis = initialTarget - initialPosition;
+    zAxis.normalize();
+
+    yAxis = zAxis.crossProduct(xAxis);
+    yAxis.normalize();
+
+    initialTransformation(0,0) = xAxis.X;
+    initialTransformation(0,1) = xAxis.Y;
+    initialTransformation(0,2) = xAxis.Z;
+
+    initialTransformation(1,0) = yAxis.X;
+    initialTransformation(1,1) = yAxis.Y;
+    initialTransformation(1,2) = yAxis.Z;
+
+    initialTransformation(2,0) = zAxis.X;
+    initialTransformation(2,1) = zAxis.Y;
+    initialTransformation(2,2) = zAxis.Z;
```

Comment:
Could you just document this row-major order stacking for the axis vectors and why we use this ordering and the difference from column-major ordering? (rotation and its inverse)

Reply:
I have added the relevant comment in https://github.com/robotology/idyntree/pull/802/commits/e20efdaca58282ce3c16895e871fd369b84e50ce

https://github.com/robotology/idyntree/blob/e20efdaca58282ce3c16895e871fd369b84e50ce/src/visualization/src/CameraAnimator.cpp#L172-L180

Reply:
Thank you. Very clear!

Reply:
Indeed, that transpose caused me a lot of thinking in the first place. Initially, it was not transposed, but the visualization was clearly wrong. So I thought I was doing some mistakes in the interpretation, I tried the transpose and everything went fine. Later on, when I was debugging a problem with the ``stl`` meshes being in the wrong position, I discovered this thing about ``irrlicht`` using left-handed frames, and then everything made sense :sweat_smile: